### PR TITLE
Add the donner::gpu shader IR core and WGSL-rule layout engine

### PR DESCRIPTION
**Pull request by jwmcglynn**
Adds the typed shader IR core and WGSL-rule layout engine under `donner/gpu/shader/`, packet 4 of the change sequence in docs/design_docs/0053-native_gpu_hal.md (stacked on the RHI core PR). No emitters yet; those are the next packets. Nothing consumes this in production and no dependencies change.

What is in the PR:

- A typed, immutable IR scoped to exactly what the solid-fill pipeline family (`slug_fill.wgsl`) requires: scalars, vectors, `mat4x4f`, sized and runtime arrays, named structs, `texture_2d<f32>`, and samplers, with value equality and validating factory functions.
- A layout engine implementing the WGSL specification's memory layout rules per address space, including the uniform-space constraints (16-byte array strides with a `paddedFromNatural` flag documenting the emitter obligation, struct-member 16-byte alignment, and follower rounding after struct-typed members). Public APIs expose offsets, sizes, and strides for host-shareable layout metadata.
- Layout anchor tests derived from the WGSL rules that reproduce the byte layouts the existing renderer's mirror structs assert: the 288-byte solid-fill uniform block with every documented field offset, the 32-byte band record, and the 32-byte instance transform.
- `ModuleBuilder`/`FunctionBuilder` with fail-closed, type-checked-at-construction expressions and statements: binding uniqueness, stage IO validation (locations, builtins including the fragment position input), lvalue rules, block scoping (names leave scope at `endIf`/`endFor`), structural missing-return detection, and exactly the fourteen builtin functions the solid-fill family uses. Ill-typed programs return labeled `ShaderResult` errors; nothing asserts on input.
- Deterministic declaration-order text serialization of modules for golden tests; byte-identical across runs, no pointers.
- 60 model-level tests: WGSL layout corner cases (vec3 tail packing, uniform vs storage array strides, nested structs with follower rounding), builder accept/reject per rule, and serialization determinism with an embedded golden.

The library depends only on `donner/base` (no RHI runtime dependency); the compute entry point seam is documented for the later filter-engine migration packets.

Validation: `bazel test //...` green locally (865 targets, 0 failures; `shader_tests` 60/60 plus the RHI core's 118), `bazel build //donner/...` green, `gen_cmakelists.py --check` passes, zero WebGPU wrapper tokens (grep-verified), clang-format/buildifier clean. The diff passed an independent sub-agent review (two blocking findings fixed with committed red-to-green sequences: a missed WGSL uniform follower-rounding rule and the missing fragment position builtin) and a privacy sweep before opening.

### CODE REVIEW

File: donner/gpu/shader/IrModule.cc [hunk 1]
```diff
@@ -0,0 +1,808 @@
...
+    return fail(Err(std::format("return type mismatch: function returns {}, got {}",
+                                function_.returnType ? function_.returnType->toString() : "void",
+                                value.type().toString()),
+                    function_.name));
+  }
+
+  IrStmt::Data data;
+  data.kind = IrStmt::Kind::Return;
+  data.exprs = {value};
+  append(IrStmt(std::move(data)));
```
[the remark this answers is not in the document]
Comment (jwmcglynn) [reply donner/gpu/shader/IrModule.cc:540]: 🤖 Fixed in 8d8076b72 (red-first tests in 533d2103d): return expressions are scope-checked via the shared ref re-verification, fwidth/textureSample are rejected in vertex-reachable functions via a per-function fragment-only-builtin mask, the for-update clause can no longer reference body-block locals, and location+builtin on one IO is rejected.

File: donner/gpu/shader/IrExpr.cc [hunk 2]
```diff
@@ -0,0 +1,748 @@
...
+    case BuiltinFn::Fract:
+    case BuiltinFn::Sqrt:
+    case BuiltinFn::Fwidth:
+    case BuiltinFn::Round:
+      if (args.size() != 1) return argCountError(1);
+      if (!args[0].type().isFloatScalarOrVector()) {
+        return ShaderError{
+            std::format("builtin requires f32 scalar or vector, got {}", TypeName(args[0])), label};
+      }
+      return args[0].type();
```
[the remark this answers is not in the document]
Comment (jwmcglynn) [reply donner/gpu/shader/IrExpr.cc:655]: 🤖 Fixed in 8d8076b72 (red-first tests in 533d2103d): return expressions are scope-checked via the shared ref re-verification, fwidth/textureSample are rejected in vertex-reachable functions via a per-function fragment-only-builtin mask, the for-update clause can no longer reference body-block locals, and location+builtin on one IO is rejected.

File: donner/gpu/shader/IrModule.cc [hunk 3]
```diff
@@ -0,0 +1,808 @@
...
+ShaderStatus FunctionBuilder::forContinuing(const IrExpr& lhs, const IrExpr& rhs) {
+  if (std::optional<ShaderError> error = checkUsable()) {
+    return *error;
+  }
+  if (blockStack_.empty() || blockStack_.back().kind != BlockFrame::Kind::ForBody) {
+    return fail(Err("forContinuing without an open for", function_.name));
+  }
+  if (!lhs.isMutableLvalue() || !(lhs.type() == rhs.type())) {
+    return fail(Err("for continuing must be a type-correct assignment to a mutable lvalue",
+                    function_.name));
```
[the remark this answers is not in the document]
Comment (jwmcglynn) [reply donner/gpu/shader/IrModule.cc:425]: 🤖 Fixed in 8d8076b72 (red-first tests in 533d2103d): return expressions are scope-checked via the shared ref re-verification, fwidth/textureSample are rejected in vertex-reachable functions via a per-function fragment-only-builtin mask, the for-update clause can no longer reference body-block locals, and location+builtin on one IO is rejected.

File: donner/gpu/shader/IrModule.cc [hunk 4]
```diff
@@ -0,0 +1,808 @@
...
+    if (param.builtin) {
+      if (*param.builtin != BuiltinInput::InstanceIndex || !(param.type == IrType::U32())) {
+        return ShaderError{
+            std::format("vertex builtin input {} must be instance_index: u32", param.name.str()),
+            name};
+      }
+    } else if (!param.location) {
+      return ShaderError{
+          std::format("vertex input {} needs a location or builtin", param.name.str()), name};
+    }
```
[the remark this answers is not in the document]
Comment (jwmcglynn) [reply donner/gpu/shader/IrModule.cc:830]: 🤖 Fixed in 8d8076b72 (red-first tests in 533d2103d): return expressions are scope-checked via the shared ref re-verification, fwidth/textureSample are rejected in vertex-reachable functions via a per-function fragment-only-builtin mask, the for-update clause can no longer reference body-block locals, and location+builtin on one IO is rejected.
